### ISSUE TITLE
ErrorException in Builder.php line 2451: Call to undefined method Illuminate\Database\Query\Builder::hasRole()

### ISSUE THREAD

**Issue by jmichaelterenin** [santigarcor/laratrust#134]
Hi there,
First off, thank you for providing an extension of Zizaco's Entrust for Laravel 5.0+.

I am having problems with an upgrade from 4.2 to 5.4. I have to make the jump to 5.4 because of several packages that I'm using that do not work in lower versions of Laravel 5. It has not been an issue (making the jump across multiple versions) thus far.

I'm running into an error whereby the hasRole() method is not being recognized by the User model. 
Here is the line of code it is failing on (user is logged in):
`if (\Auth::check() && (\Auth::user()->hasRole('Provider') || \Auth::user()->hasRole('Data Entry'))) {`

Here is the User model:
`class User extends Model
{
	use LaratrustUserTrait, SyncableGraphNodeTrait, Messagable, RevisionableTrait;`

 Not sure if you would need the entire class, let me know.
Is it an issue that I am NOT using the User model that extends: Authenticatable? I wouldn't think so, as I previously had the User model extending Model in 4.2

One last piece of information, here is what is in auth.php:

```
   'providers' => [
        'users' => [
           'driver' => 'eloquent',
            'model' => App\User::class,
        ],

        // 'users' => [
        //     'driver' => 'database',
        //     'table' => 'users',
        // ],
    ],

```
I tried changing to the database driver, but this provided different errors, and I don't think it's what I want anyway, but I'm obviously not sure, since I'm seeking help.

**Comment by KKSzymanowski** [comment]
This error message:
> ErrorException in Builder.php line 2451: Call to undefined method Illuminate\Database\Query\Builder::hasRole()

suggests you did something with the query builder before calling `hasRole`. For example: `User::where('name', 'foo')->hasRole()`. The `where` call will instantiate the query builder and return it causing this code to fail.

Are you sure, the provided code:
```php
if (\Auth::check() && (\Auth::user()->hasRole('Provider') || \Auth::user()->hasRole('Data Entry'))) {
```
is the part causing the error?

To see if the method really exists you can for example run `php artisan tinker` in your console, then run
```
\Namespace\Of\User::first()->hasRole('Provider');
```

**Comment by jmichaelterenin** [comment]
BadMethodCallException with message 'Call to undefined method Illuminate\Database\Query\Builder::hasRole()' error when running:
`\App\User::first()->hasRole('Provider')`

**Comment by KKSzymanowski** [comment]
That's peculiar. Did you by any change overwrite the `first` method? It should always return an Eloquent model or `null`.

**Comment by jmichaelterenin** [comment]
No sir. \App\user::first() does return the first record in the database table.

**Comment by jmichaelterenin** [comment]
I just tried this in a test controller:
`$usrRepo = \App::make('\App\UserRepository');
    	if (($usrRec = $usrRepo->recordExists(['id' => 2])) != null) {
    		echo $usrRec->hasRole('provider');	
    	}`
same result

**Comment by KKSzymanowski** [comment]
Ah sorry, I see what's going on. I forgot Eloquent tries to find a method on a query builder when it's not found on the model, should've remembered that 😄.

What's the output of:
```php
dd(
    class_uses(\App\User::class),
    get_class_methods(\App\User::first())
);
```
Maybe it will give us a hint what's going on.

Which version of Laratrust are you using?

**Comment by jmichaelterenin** [comment]
`array:1 [▼
  "Illuminate\Notifications\Notifiable" => "Illuminate\Notifications\Notifiable"
]
array:179 [▼
  0 => "__construct"
  1 => "clearBootedModels"
  2 => "fill"
  3 => "forceFill"
  4 => "newInstance"
  5 => "newFromBuilder"
  6 => "on"
  7 => "onWriteConnection"
  8 => "all"
  9 => "with"
  10 => "load"
  11 => "update"
  12 => "push"
  13 => "save"
  14 => "saveOrFail"
  15 => "destroy"
  16 => "delete"
  17 => "forceDelete"
  18 => "query"
  19 => "newQuery"
  20 => "newQueryWithoutScopes"
  21 => "newQueryWithoutScope"
  22 => "newEloquentBuilder"
  23 => "newCollection"
  24 => "newPivot"
  25 => "toArray"
  26 => "toJson"
  27 => "jsonSerialize"
  28 => "fresh"
  29 => "replicate"
  30 => "is"
  31 => "getConnection"
  32 => "getConnectionName"
  33 => "setConnection"
  34 => "resolveConnection"
  35 => "getConnectionResolver"
  36 => "setConnectionResolver"
  37 => "unsetConnectionResolver"
  38 => "getTable"
  39 => "setTable"
  40 => "getKeyName"
  41 => "setKeyName"
  42 => "getQualifiedKeyName"
  43 => "getKeyType"
  44 => "setKeyType"
  45 => "getIncrementing"
  46 => "setIncrementing"
  47 => "getKey"
  48 => "getQueueableId"
  49 => "getRouteKey"
  50 => "getRouteKeyName"
  51 => "getForeignKey"
  52 => "getPerPage"
  53 => "setPerPage"
  54 => "__get"
  55 => "__set"
  56 => "offsetExists"
  57 => "offsetGet"
  58 => "offsetSet"
  59 => "offsetUnset"
  60 => "__isset"
  61 => "__unset"
  62 => "__call"
  63 => "__callStatic"
  64 => "__toString"
  65 => "__wakeup"
  66 => "attributesToArray"
  67 => "relationsToArray"
  68 => "getAttribute"
  69 => "getAttributeValue"
  70 => "getRelationValue"
  71 => "hasGetMutator"
  72 => "setAttribute"
  73 => "hasSetMutator"
  74 => "fillJsonAttribute"
  75 => "fromJson"
  76 => "fromDateTime"
  77 => "getDates"
  78 => "setDateFormat"
  79 => "hasCast"
  80 => "getCasts"
  81 => "getAttributes"
  82 => "setRawAttributes"
  83 => "getOriginal"
  84 => "syncOriginal"
  85 => "syncOriginalAttribute"
  86 => "isDirty"
  87 => "isClean"
  88 => "getDirty"
  89 => "append"
  90 => "setAppends"
  91 => "getMutatedAttributes"
  92 => "cacheMutatedAttributes"
  93 => "observe"
  94 => "getObservableEvents"
  95 => "setObservableEvents"
  96 => "addObservableEvents"
  97 => "removeObservableEvents"
  98 => "saving"
  99 => "saved"
  100 => "updating"
  101 => "updated"
  102 => "creating"
  103 => "created"
  104 => "deleting"
  105 => "deleted"
  106 => "flushEventListeners"
  107 => "getEventDispatcher"
  108 => "setEventDispatcher"
  109 => "unsetEventDispatcher"
  110 => "addGlobalScope"
  111 => "hasGlobalScope"
  112 => "getGlobalScope"
  113 => "getGlobalScopes"
  114 => "hasOne"
  115 => "morphOne"
  116 => "belongsTo"
  117 => "morphTo"
  118 => "getActualClassNameForMorph"
  119 => "hasMany"
  120 => "hasManyThrough"
  121 => "morphMany"
  122 => "belongsToMany"
  123 => "morphToMany"
  124 => "morphedByMany"
  125 => "joiningTable"
  126 => "touches"
  127 => "touchOwners"
  128 => "getMorphClass"
  129 => "getRelations"
  130 => "getRelation"
  131 => "relationLoaded"
  132 => "setRelation"
  133 => "setRelations"
  134 => "getTouchedRelations"
  135 => "setTouchedRelations"
  136 => "touch"
  137 => "setCreatedAt"
  138 => "setUpdatedAt"
  139 => "freshTimestamp"
  140 => "freshTimestampString"
  141 => "usesTimestamps"
  142 => "getCreatedAtColumn"
  143 => "getUpdatedAtColumn"
  144 => "getHidden"
  145 => "setHidden"
  146 => "addHidden"
  147 => "getVisible"
  148 => "setVisible"
  149 => "addVisible"
  150 => "makeVisible"
  151 => "makeHidden"
  152 => "getFillable"
  153 => "fillable"
  154 => "getGuarded"
  155 => "guard"
  156 => "unguard"
  157 => "reguard"
  158 => "isUnguarded"
  159 => "unguarded"
  160 => "isFillable"
  161 => "isGuarded"
  162 => "totallyGuarded"
  163 => "getAuthIdentifierName"
  164 => "getAuthIdentifier"
  165 => "getAuthPassword"
  166 => "getRememberToken"
  167 => "setRememberToken"
  168 => "getRememberTokenName"
  169 => "can"
  170 => "getEmailForPasswordReset"
  171 => "sendPasswordResetNotification"
  172 => "cant"
  173 => "cannot"
  174 => "notifications"
  175 => "readNotifications"
  176 => "unreadNotifications"
  177 => "notify"
  178 => "routeNotificationFor"
]`

Laratrust 3.2

**Comment by KKSzymanowski** [comment]
There:
```
array:1 [▼ "Illuminate\Notifications\Notifiable" => "Illuminate\Notifications\Notifiable" ]
```
Your User class doesn't use the `LaratrustUserTrait`(nor `SyncableGraphNodeTrait`, `Messagable`, `RevisionableTrait`). Please verify your trait imports.

**Comment by KKSzymanowski** [comment]
> Is there a better way to display this in the editor? Nvm

Use triple accent signs for big code chunks. See [Github Flavored Markdown](https://guides.github.com/features/mastering-markdown/) for more(specifically the Syntax highlighting section).

```
array:1 [▼
"Illuminate\Notifications\Notifiable" => "Illuminate\Notifications\Notifiable"
]
array:179 [▼
0 => "__construct"
1 => "clearBootedModels"
2 => "fill"
3 => "forceFill"
4 => "newInstance"
...
```

**Comment by jmichaelterenin** [comment]
Thanks so much KK. Turns out I had renamed the original User model file from the Laravel 5.4 install, but left it in the same directory, and composer was finding it before it found my revised one.

I have a different error now, related to not having the User model being extended from Authenticatable, but I'll forge forward, thanks. If you have any tips on:

Call to undefined method Illuminate\Database\Query\Builder::getAuthIdentifierName()

I'll take it. I mean I know I need for the User model to be extended from Authenticatable to make maximum use of Laravel's Authentication features (and avoid using a 3rd party package like Zizaco Confide), but I just haven't got there yet.

Thanks again!

**Comment by KKSzymanowski** [comment]
> Call to undefined method Illuminate\Database\Query\Builder::getAuthIdentifierName()

Yup, that's a method from [Authenticatable trait](https://github.com/laravel/framework/blob/5.4/src/Illuminate/Auth/Authenticatable.php#L19).
I'm afraid, with Laratrust 3.2 you can't **extend** Authenticatable, you need to **use** the trait because of `can` method conflicts. [Here](https://github.com/santigarcor/laratrust/issues/90) is the full story and this is the solution I use in one of my projects:
```php
use Illuminate\Auth\Authenticatable;
use Laratrust\Traits\LaratrustUserTrait;
use Illuminate\Notifications\Notifiable;
use Illuminate\Auth\Passwords\CanResetPassword;
use Illuminate\Foundation\Auth\Access\Authorizable;
use Illuminate\Contracts\Auth\Authenticatable as AuthenticatableContract;
use Illuminate\Contracts\Auth\Access\Authorizable as AuthorizableContract;
use Illuminate\Contracts\Auth\CanResetPassword as CanResetPasswordContract;

class User extends Model implements AuthenticatableContract, AuthorizableContract, CanResetPasswordContract
{

    use Authenticatable,
        Authorizable,
        CanResetPassword,
        Notifiable,
        LaratrustUserTrait {
        LaratrustUserTrait::can insteadof Authorizable;
        Authorizable::can as authorizableCan;
    }
```
Basically you need to tell PHP to use a method from one trait and rename the method from the other.

**Comment by jmichaelterenin** [comment]
Oh wow, I am very glad that you supplied this to me, thanks again, and I will use this.
Thanks for the great work!
